### PR TITLE
Permissions constraints

### LINKED ISSUE #593: Permission popup: add constraints
## Overview
Continue the work start on the permission popup: 

Some permissions are depending on other permissions, for instance, you cannot give "can_grant_view:content" to a user without giving him "can_view:content" first.

Implement them all in a proper way (readable code).

image

## Acceptance criteria
- [ ] Implement the rules described in the [devdoc perm page](https://france-ioi.github.io/algorea-devdoc/design/access-rights/items/#changing-and-propagating-permissions), in the "Constraint on “receiver”" column.
In practice, in the UI:
- [ ] do not let the user reduce the "required" permission, by greying out the lower option (and its icon on the left) and displaying an explanation tooltip on hover
- [ ] do not let the user increases the "requiring" permission,  by greying out the lower option (and its icon on the left) and displaying an explanation tooltip on hover

### PR DESCRIPTION
**Pull request by GabrielVidal1**
## Description

This is a redo of PR #692 

Fixes #593 

Permissions constraints are added through a custom validator in the permissions edition dialog box as suggested [here](https://github.com/France-ioi/AlgoreaFrontend/pull/692#issuecomment-925890649).
The values that are invalid due to permissions constraints are not grayed out according to the [constraints on receiver](https://france-ioi.github.io/algorea-devdoc/design/access-rights/items/#changing-and-propagating-permissions). (for the `progress-select` and the `switch-field`)

### Testing

I wrote some unit tests for the permissions constraints custom validators [here](https://github.com/France-ioi/AlgoreaFrontend/pull/881/files#diff-04f985d45d516233c178c9906c8aa71307bcd517d4977add1d41551486b229f8)
For now every valid and invalid combinations of permissions are tested for the fields 'Can view' and 'Can grant view'.
Should I write the tests for the rest (the 4 other fields) ? or testing every cases is not useful ?

## Test cases

- [x] **Case 1**: You have the rights to edit the permissions of a user
  1. Given I am the user arbonenfant
  2. Start observing the groupe [UE Info - 2019-2020](https://dev.algorea.org/branch/permissions-constraints/en/#/groups/by-id/472375073987524072;path=/details)
  3. Go to the item [SSGeom](https://dev.algorea.org/branch/permissions-constraints/en/#/activities/by-id/1787514136807846970;path=4702,4102,1012286565380301759;attempId=0/details/progress/chapter) and in the Progress tab > Chapter View > Users, open the permission edit dialogue box for any users
  4. You can now edit permission (as an owner of the item, so almost all permissions)
  5. You can try changing permissions that depend on others, for example toggling 'Can attach official sessions' on and setting 'Can View' to 'Nothing' to invalidate the permissions

- [x] **Case 2**: You have the rights to edit the permissions of a user, but **not all permissions**
  1. Given I am the user arbonenfant
  2. Start observing the groupe [UE Info - 2019-2020](https://dev.algorea.org/branch/permissions-constraints/en/#/groups/by-id/472375073987524072;path=/details)
  3. Go to the item [SNT](https://dev.algorea.org/branch/permissions-constraints/en/#/activities/by-id/1352246428241737349;path=4702;attempId=0/details/progress/chapter) and in the Progress tab > Chapter View > Users, open the permission edit dialogue box for any users
  4. You can edit the permissions, but are very limited so you can see the permissions constraints at work.

### CODE REVIEW

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html [hunk 1]
```diff
@@ -22,12 +22,16 @@
...
+          i18n-title title="Can view" icon="fa fa-eye"
+          [collapsible]="!form.errors?.canView"
+        >
           <ng-template #content let-collapsed>
             <alg-progress-select
               [collapsed]="collapsed"
               formControlName="canView"
-              [values]="canViewValues"
+              [values]="progressSelectValues.canViewValues"
+              [validationErrors]="form.errors?.canView"
```
Comment (GabrielVidal1) [inline src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html:34]: In the PR #692, @ThomasJuster commented ([link of the thread here](https://github.com/France-ioi/AlgoreaFrontend/pull/692#discussion_r737474645))
> I don't understand the link with validation, to me validation is defined at form-level and errors are stored into `form.errors` object, into properties that may match input names
> 
> Here, input name and error property seem to match, so according to the [doc](https://angular.io/api/forms/AbstractControl#errors) we should be able to also access the error from the `canView` FormControl via property `errors`

I could access the form control form inside the component, but after some testing, the form group errors are not mapped to the form controls with the same name, so the only way to access the error, is by getting it from the parent.
Also I couldn't manage to display the errors after a change in the form, as the fields have no way of knowing the form has changed.

**Comment by ThomasJuster** [comment]
I didn't view it at first, the error message is shady
image
I think it should be more something like `"All with grant" (or This level of permission) can be granted only if "Can view" is at least "Content"`
@smadbe do you have any specific error message in mind ?
Plus, same as my [last comment](https://github.com/France-ioi/AlgoreaFrontend/pull/881#discussion_r796827493), it could (should ?) be treated in another PR.

**Comment by ThomasJuster** [comment]
I think tooltip + a title attribute is redundant :smile: 
Algorea screenshot permissions

File: src/app/shared/http-services/group-permissions.service.ts [hunk 2]
```diff
@@ -29,24 +42,28 @@ export class GroupPermissionsService {
 
   constructor(private http: HttpClient) { }
 
-  getPermissions(sourceGroupId: string, groupId: string, itemId: string): Observable<GroupPermissions> {
+  getPermissions(sourceGroupId: string, groupId: string, itemId: string): Observable<GroupPermissionsInfo> {
     return this.http
-      .get<unknown>(`${appConfig.apiUrl}/groups/${sourceGroupId}/permissions/${groupId}/${itemId}`)
-      .pipe(
-        decodeSnakeCase(groupPermissionsDecoder),
+      .get<GroupPermissionsInfo>(`${appConfig.apiUrl}/groups/${sourceGroupId}/permissions/${groupId}/${itemId}`).pipe(
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/shared/http-services/group-permissions.service.ts:47]: fixed

File: src/app/shared/http-services/group-permissions.service.ts [hunk 3]
```diff
@@ -29,24 +42,28 @@ export class GroupPermissionsService {
...
+
+    const body = {
+      can_view: permissions.canView,
+      can_grant_view: permissions.canGrantView,
+      can_watch: permissions.canWatch,
+      can_edit: permissions.canEdit,
+      can_make_session_official: permissions.canMakeSessionOfficial,
+      is_owner: permissions.isOwner,
+      can_enter_from: undefined,
+      can_enter_until: undefined,
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/shared/http-services/group-permissions.service.ts:63]: fixed

File: src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.ts [hunk 4]
```diff
@@ -66,6 +65,8 @@ export class GroupProgressGridComponent implements OnChanges, OnDestroy {
         canEdit: 'none',
         canMakeSessionOfficial: false,
         isOwner: true,
+        canEnterFrom: new Date(),
+        canEnterUntil: new Date(),
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.ts:69]: fixed

File: src/app/modules/item/helpers/item-permissions-constraints.ts [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+  }
+
+  if (receiverPermissions.canEdit === 'all_with_grant') {
+    if (!giverPermissions.isOwner) {
+      errors.push($localize`You need to be owner of this item`);
+    }
+  } else {
+
+    // if receiverPermissions.canEdit is 'children' or 'all_with_grant'
+    if (!(giverPermissions.canEdit === 'all_with_grant')) {
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.ts:153]: fixed

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 6]
```diff
@@ -37,43 +50,46 @@ export class PermissionsEditDialogComponent implements OnChanges {
...
-    }
+  constructor(private fb: FormBuilder) {
+    combineLatest([ this.form.valueChanges, this.regenerateValues.asObservable() ])
+      .subscribe(() => {
+        if (this.permissions && this.giverPermissions) {
+          const receiverPermissions = this.form.value as GroupPermissions;
+          this.permissionsDialogData = generateValues(this.targetType, receiverPermissions, this.giverPermissions);
+        }
+      });
+  }
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts:63]: fixed

File: src/app/modules/item/helpers/item-permissions-constraints.spec.ts [hunk 7]
```diff
@@ -0,0 +1,346 @@
+import { canGrantViewValues, canViewValues } from './item-permissions';
+import { validateCanGrantView, validateCanView } from './item-permissions-constraints';
+
+function flatMap<T>(arr: T[][]) {
+  return arr.reduce((acc, val) => acc.concat(val), []);
+}
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.spec.ts:6]: I added tests for the rest (the validation function for the permissions `can_edit`, `can_watch`, `can_make_session_official` and `is_owner`)

File: src/app/modules/item/helpers/item-permissions-constraints.ts [hunk 8]
```diff
@@ -0,0 +1,263 @@
+import { ValidationErrors, ValidatorFn, AbstractControl } from '@angular/forms';
+import { GroupPermissions } from 'src/app/shared/http-services/group-permissions.service';
+import { PermissionsInfo, canGrantViewValues, canViewValues, permissionsInfoString } from './item-permissions';
+import { TypeFilter } from '../components/composition-filter/composition-filter.component';
+import { PermissionsDialogData } from '../components/permissions-edit-dialog/permissions-edit-dialog.component';
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.ts:5]: fixed, I moved the interface and the function using it to the file `permissions-edit-dialog-texts.ts` which already contains functions and logic used by the component

File: src/app/modules/item/helpers/item-permissions-constraints.ts [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+    // if receiverPermissions.canEdit is 'children' or 'all_with_grant'
+    if (!(giverPermissions.canEdit === 'all_with_grant')) {
+      errors.push(`You need ${bolden(permissionsInfoString.canEdit.string)} to be ${
+        bolden(permissionsInfoString.canEdit.all_with_grant)}`);
+    }
+  }
+
+  return errors.length === 0 ? {} : { canEdit: errors };
+}
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.ts:160]: fixed

**Comment by GabrielVidal1** [comment]
I think all translation related fixes will be in another PR

**Comment by GabrielVidal1** [comment]
> I think tooltip + a title attribute is redundant 😄 Algorea screenshot permissions

This is now fixed

**Comment by GabrielVidal1** [comment]
I made an issue (#978) regarding improving the internationalization for this PR, because this PR is already big and hard to review as is.
Now let's focus on merging this one as long as everything works, and improvements will be done after.

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog-texts.ts [hunk 10]
```diff
@@ -13,125 +15,133 @@ function getTargetTypeString(targetType: TypeFilter): string {
...
+export function generateCanViewValues(
+  targetType: TypeFilter,
+): ProgressSelectValue<GroupPermissions['canView']>[]{
   const targetTypeString = getTargetTypeString(targetType);
   return [
     {
       value: 'none',
-      label: $localize`Nothing`,
+      label: permissionsInfoString.canView.none,
       comment: $localize`${targetTypeString} can\'t see the item`
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog-texts.ts:43]: I made an issue for this improvement (#978) cf [this message](https://github.com/France-ioi/AlgoreaFrontend/pull/881#issuecomment-1079951076).

File: src/app/modules/item/helpers/item-permissions.ts [hunk 11]
```diff
@@ -21,3 +26,37 @@ export function canCurrentUserViewItemContent(item: ItemPermissionsInfo): boolea
...
+    answer_with_grant: $localize`Answer and grant`,
+  },
+  canEdit: {
+    string: $localize`Can edit`,
+    none: $localize`Nothing`,
+    children: $localize`Children`,
+    all: $localize`All`,
+    all_with_grant: $localize`All and grant`,
+  }
+};
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions.ts:62]: This is a good idea, however I think this PR is already big, so I referenced this improvement in a new issue I made (#978) cf [this message](https://github.com/France-ioi/AlgoreaFrontend/pull/881#issuecomment-1079951076).

File: src/app/modules/item/helpers/item-permissions-constraints.spec.ts [hunk 12]
```diff
@@ -0,0 +1,632 @@
...
+      receiverPermissions: combinations({
+        canWatch: [ 'answer_with_grant' ] as const,
+        canView: [ 'none', 'info' ] as const,
+      }),
+      giverPermissions: combinations({
+        canWatch: canWatchValues,
+        isOwner: [ true, false ] as const,
+      }),
+    })) {
+      expect(validateCanWatch(p.receiverPermissions, p.giverPermissions)).toBeDefined();
```
Comment (ThomasJuster) [inline src/app/modules/item/helpers/item-permissions-constraints.spec.ts:446]: Isn't it `expect(validate(…).canWatch).toBeDefined()` ?
If the validate function always returns an object (empty or not), this test will always be true

Same below

File: src/app/shared/http-services/group-permissions.service.ts [hunk 13]
```diff
@@ -5,22 +5,35 @@ import { map } from 'rxjs/operators';
 import { appConfig } from '../helpers/config';
 import { assertSuccess, SimpleActionResponse } from './action-response';
 import * as D from 'io-ts/Decoder';
-import { pipe } from 'fp-ts/function';
-import { decodeSnakeCase } from 'src/app/shared/operators/decode';
-import { permissionsDecoder, Permissions } from '../helpers/group-permissions';
+import { permissionsDecoder } from 'src/app/modules/item/helpers/item-permissions';
+import { pipe } from 'fp-ts/lib/function';
```
Comment (ThomasJuster) [inline src/app/shared/http-services/group-permissions.service.ts:9]: I know the autocomplete generates an import this way, but we should import from `fp-ts/function` to let the bundler decide whether to use the esm or cjs pkg build

Comment (GabrielVidal1) [reply src/app/shared/http-services/group-permissions.service.ts]: fixed

File: src/app/modules/item/helpers/item-permissions-constraints.spec.ts (hunk 12 above)
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.spec.ts]: You're right, its fixed now

File: src/app/modules/item/helpers/item-permissions-constraints.ts [hunk 14]
```diff
@@ -0,0 +1,221 @@
+import { ValidationErrors, ValidatorFn, AbstractControl } from '@angular/forms';
+import { GroupPermissions } from 'src/app/shared/http-services/group-permissions.service';
+import { PermissionsInfo, canGrantViewValues, canViewValues, permissionsInfoString } from './item-permissions';
+
+const bolden = (text: string): string => `<b>${text}</b>`;
```
Comment (smadbe) [inline src/app/modules/item/helpers/item-permissions-constraints.ts:5]: IMO, this file is violating a bit software engineering rules by mixing roles:
1. it finds constraint violations in permissions
2. it convert these violations in texts
3. in format it (this bold)

This should be done at 3 different places. Actually 1+2 is still ok, you can maybe leave it as it is, but not 3.

File: src/app/modules/shared-components/components/collapsible-section/switch-field/switch-field.component.ts [hunk 15]
```diff
@@ -32,6 +32,10 @@ export class SwitchFieldComponent implements ControlValueAccessor {
   @Input() value = false;
   @Input() collapsed = false;
 
+  // If this is not empty, the switch will be disabled, and the content of this array will appear as a tooltip
```
Comment (smadbe) [inline src/app/modules/shared-components/components/collapsible-section/switch-field/switch-field.component.ts:35]: Is it useful to have them optional? It means `[]` has a different meaning from `undefined`.

**Review by smadbe — CHANGES_REQUESTED** [review]
Great work. I guess it was not easy.

First of all, a visual review: 
- when view is squeezed (narrow screen), it seems the the size of the red text does not display in the same font size as the other texts, I see no reason for that
Screenshot 2022-04-05 at 17 39 54.
BTW, I would use tag instead so that it is more visible. So, something like:
Screenshot 2022-04-06 at 10 06 31
(For info, for the demo, I've added `<p-tag styleClass="mr-2" severity="danger" value='This user needs "Can view" to be at least "Content"' style="padding-right: 10px;">`)
 - the rest looks good :-)  

Of course, texts will have to rephrased. It can be done in a separate PR. 👍 

For the code, I have not rechecked everything in details but my few remarks are in the code.

File: src/app/modules/item/helpers/item-permissions-constraints.ts (hunk 14 above)
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.ts]: fixed, the `validate...()` functions now return an object describing the constraint that is not met. This object is latter transformed into a readable error message

File: src/app/modules/shared-components/components/collapsible-section/switch-field/switch-field.component.ts (hunk 15 above)
Comment (GabrielVidal1) [reply src/app/modules/shared-components/components/collapsible-section/switch-field/switch-field.component.ts:35]: fixed, now the array is empty by default

**Comment by GabrielVidal1** [comment]
> Great work. I guess it was not easy.
> 
> First of all, a visual review:
> 
> * when view is squeezed (narrow screen), it seems the the size of the red text does not display in the same font size as the other texts, I see no reason for that
>   Screenshot 2022-04-05 at 17 39 54.
>   BTW, I would use tag instead so that it is more visible. So, something like:
>   Screenshot 2022-04-06 at 10 06 31
>   (For info, for the demo, I've added `<p-tag styleClass="mr-2" severity="danger" value='This user needs "Can view" to be at least "Content"' style="padding-right: 10px;">`)
> * the rest looks good :-)

Now error messages are shown in a tag in the header of the collapsible section

The issue with the tag is that sometimes there can be 2 error messages, because two constraints are not met (one for the giver and one for the reciever), but in practice, this is very rare case, so if there is more than one error message, only the first one is shown.

File: src/app/modules/item/helpers/item-permissions-constraints.ts [hunk 16]
```diff
@@ -1,8 +1,29 @@
...
+    expectedValue: ConstrainedPermissions[Permission],
+    on: 'giver' | 'receiver',
+    constraintType?: ConstrainedPermissions[Permission] extends boolean ? never : 'atLeast',
+  ): ConstraintError<Permission> => ({
+    permission, expectedValue, constraintType: constraintType ?? 'equals', on
+  });
+}
 
-const bolden = (text: string): string => `<b>${text}</b>`;
+// const bolden = (text: string): string => `<b>${text}</b>`;
```
Comment (smadbe) [inline src/app/modules/item/helpers/item-permissions-constraints.ts:26]: you keep this commented command?

**Review by smadbe — COMMENTED** [review]
I think it can be merged as it is and then we'll do improvements based on this.

**Review by smadbe — APPROVED** [review]

File: src/app/modules/item/helpers/item-permissions-constraints.ts (hunk 16 above)
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.ts]: fixed

**Pull request merged**